archive/tar_stream.ts

```diff
@@ -0,0 +1,450 @@
+   *
+   * @example Usage
+   * ```ts no-eval
+   * import { TarStream } from "@std/archive/tar-stream";
+   *
+   * await ReadableStream.from([
+   *   {
+   *     path: 'potato/'
+   *   },
+   *   {
+   *     path: 'deno.json',
+   *     size: (await Deno.stat('deno.json')).size,
+   *     iterable: (await Deno.open('deno.json')).readable
+   *   },
+   *   {
+   *     path: '.vscode/settings.json',
+   *     size: (await Deno.stat('.vscode/settings.json')).size,
+   *     iterable: (await Deno.open('.vscode/settings.json')).readable
+   *   }
+   * ])
+   *   .pipeThrough(new TarStream())
+   *   .pipeThrough(new CompressionStream('gzip'))
+   *   .pipeTo((await Deno.create('./out.tar.gz')).writable)
+   * ```
+   */
+  get writable(): WritableStream<TarStreamInput> {
+    return this.#writable;
+  }
+}
+
+function parsePath(
+  path: string,
+): [Uint8Array, Uint8Array] {
+  const name = new TextEncoder().encode(path);
+  if (name.length <= 100) {
+    return [new Uint8Array(0), name];
+  }
+
+  if (name.length > 256) {
+    throw new RangeError("Path cannot exceed 256 bytes");
+  }
+
+  // If length of last part is > 100, then there's no possible answer to split the path
+  let suitableSlashPos = Math.max(0, name.lastIndexOf(SLASH_CODE_POINT)); // always holds position of '/'
+  if (name.length - suitableSlashPos > 100) {
+    throw new RangeError("Filename cannot exceed 100 bytes");
+  }
+
+  for (
+    let nextPos = suitableSlashPos;
+    nextPos > 0;
+    suitableSlashPos = nextPos
+  ) {
+    // disclaimer: '/' won't appear at pos 0, so nextPos always be > 0 or = -1
+    nextPos = name.lastIndexOf(SLASH_CODE_POINT, suitableSlashPos - 1);
+    // disclaimer: since name.length > 100 in this case, if nextPos = -1, name.length - nextPos will also > 100
+    if (name.length - nextPos > 100) {
+      break;
+    }
+  }
+
+  const prefix = name.slice(0, suitableSlashPos);
+  if (prefix.length > 155) {
+    throw new TypeError(
+      "Path needs to be split-able on a forward slash separator into [155, 100] bytes respectively",
+    );
+  }
+  return [prefix, name.slice(suitableSlashPos + 1)];
+}
+
+/**
+ * Asserts that the path provided is valid for a {@linkcode TarStream}.
+ *
+ * It provides a means to check that a path is valid before pipping it through
+ * the `TarStream`, where if invalid will throw an error. Ruining any progress
+ * made when archiving.
+ *
+ * @param path The path as a string
+ *
+ * @example Usage
+ * ```ts no-assert
+ * import { assertValidPath } from "@std/archive";
+ *
+ * assertValidPath('MyFile.txt');
+ * ```
+ */
+export function assertValidPath(path: string) {
+  parsePath(path);
+}
+
+/**
+ * Asserts that the options provided are valid for a {@linkcode TarStream}.
+ *
+ * @param options The TarStreamOptions
+ *
+ * @example Usage
+ * ```ts no-assert
+ * import { assertValidTarStreamOptions } from "@std/archive";
+ *
+ * assertValidTarStreamOptions({ mode: 755 })
+ * ```
```

Comment:
Ditto. Can you please show to use this during iteration? I guess we could use the same snippet for both this function and the above `assertValidPath()`, as long as the shared snippet uses both functions.